### ISSUE TITLE
[Bug] XML Variables query not working 

### ISSUE THREAD

**Issue by txjmb** [yesoreyeram/grafana-infinity-datasource#146]
**Describe the bug**
Many thanks for the fantastic plugin!

Given a valid xml document url, it does not appear to be possible to get a working variable query in Grafana using Infinity with the versions I am using.  Identical panel queries work, just not variable queries.

You should be able to reproduce with these steps:

Create new Variable Query on a new dashboard.

Use Infinity datasource, 
Type:  XML type, URL, 
URL: https://www.youtube.com/feeds/videos.xml?channel_id=UCkHr2Z0JWH8KGmQNDp4QXLw
Rows/Root: fee.entry

Add field for "title".

This query works fine in a Panel, but not in variable.  Nothing shows up in preview or in dropdown.

**Version Details:**

- Grafana version :  8.1.2
- Plugin version :  0.7.8

**Additional Details**
No obvious errors showing up, just doesn't do anything at all.

**Comment by gdiana** [comment]
Hi txjmb,

I saw this and also had problems with xml path and have found that it very much depends on the author as to the syntax one should use. In your case the url is quite complex but I managed to play around a bit and get it to work.

When the xml contains objects it starts to get vague and tricky and I have provided some examples to hopefully help you.

I used https://codebeautify.org/xmlviewer# and used the tree View as shown below and as you can see it is tricky and some items are objects (in green) and one has to get into the object to get it to display correctly in a table (which I presume you are using)

![image](https://user-images.githubusercontent.com/13445152/132422681-9a133f86-8dc8-4bac-8dbf-a117c650bd6f.png)

So using the **feed.entry** as root I too was able to get the following using the **selectors** as shown in the column headers

![image](https://user-images.githubusercontent.com/13445152/132423097-d8656c83-1ff0-4642-b269-e896db2f55dd.png)

However to access objects one has to expose all objects and then drill down so to do this I used **feed.entry[0].media:group** as root I was able to access the media group stuff as follows.

![image](https://user-images.githubusercontent.com/13445152/132423331-ac1259eb-a477-4726-a5a9-88e03d098f24.png)

and

![image](https://user-images.githubusercontent.com/13445152/132423456-10315591-3719-4588-a525-3083862540c3.png)


Here using selector media.thumbnail returns [object Object]

and using elector **media.thumbnail[0]** returns {"$":{"url":"https://i2.ytimg.com/vi/5BeqlhkhRrw/hqdefault.jpg","width":"480","height":"360"}}

and then using selector **media.thumbnail[0].$**  returns {"url":"https://i2.ytimg.com/vi/5BeqlhkhRrw/hqdefault.jpg","width":"480","height":"360"}

Which is the underlying object

and finally usinbg selector **media.thumbnail[0].$.url** returns 
{"url":"https://i2.ytimg.com/vi/5BeqlhkhRrw/hqdefault.jpg","width":"480","height":"360"}

You may then replace url with width and height

So as may be seen the trick is to use the correct root and selector and my experience is one has to do some head bashing as each system may use a different tree structure and the trick is how to decode this?

**Comment by txjmb** [comment]
Thank you for the detailed response.  It's helpful to have more examples of using the selector.  The regular panel queries work well.  However, the variables query is broken.  I saw the other bug, but the behavior description was somewhat different.  I can close this one if it is definitely a duplication.

**Comment by yesoreyeram** [comment]
@gdiana v0.8.0-dev.6 release should fix this. temp demo page https://grafana-infinity-datasource.herokuapp.com/d/FpHVrG57k/191-variables-not-working?orgId=1

Lets know if it is not working as expected

Download : https://github.com/yesoreyeram/grafana-infinity-datasource/releases/tag/v0.8.0-dev.6

**Comment by gdiana** [comment]
Ok I downloaded and tried. I am using latest version of Grafana.

![image](https://user-images.githubusercontent.com/13445152/140959309-ec6d9ea5-fe15-435a-902e-5b1a4e95c354.png)

When I am using the older version (0.7.10) I get 

![image](https://user-images.githubusercontent.com/13445152/140959784-9ca74a0e-81e6-4748-a66b-efa598cf2eb5.png)

When I load your latest version I get

![image](https://user-images.githubusercontent.com/13445152/140960746-070e6152-ee9c-4e07-8762-57d8066bec03.png)

As you can see some items are not being returned using the latest version.

The SIM Information is fine but is uses GraphQL, so this appears all good. But the XML queries are not fully functional.

This was just a quick test but it appears normal queries are being affected. I will play around and see if it is a syntax error and get back.

In short it is not working!

**Comment by gdiana** [comment]
Ok I tried to get the table correct and found that all the missing variables were defined as type number in grafana. I tried changing them to string and they all appeared.

It appears there is a bug in being able to recognise numbers.

For example for Serial Number I previously had  **61110355**  whereas now I have **61110355 :**

So all NUMBERS are not being recognized except when I define them as string and which adds a colon (:) as you can see.


![image](https://user-images.githubusercontent.com/13445152/140964116-04fbc7eb-519a-46d8-84c7-fcb48b5567a4.png)

**Comment by yesoreyeram** [comment]
Thanks @gdiana for trying the new version and sorry it broke few things in your existing dashboard. But we can fix. :)

Do you mind creating an inline data version of your bug [here](https://grafana-infinity-datasource.herokuapp.com/explore?orgId=1&left=%5B%22now-1h%22,%22now%22,%22Infinity%22,%7B%22refId%22:%22A%22,%22type%22:%22json%22,%22source%22:%22url%22,%22format%22:%22table%22,%22url%22:%22https:%2F%2Fjsonplaceholder.typicode.com%2Fusers%22,%22url_options%22:%7B%22method%22:%22GET%22,%22data%22:%22%22%7D,%22root_selector%22:%22%22,%22columns%22:%5B%5D,%22filters%22:%5B%5D,%22global_query_id%22:%22%22%7D%5D) and share the URL? that will help me to understand the issue.

Shall I assume your number column have special character ":"  is the issue? But I wonder how that used to work before :)

**Comment by yesoreyeram** [comment]
BTW, the new version solves the original XML variable issue?

**Comment by gdiana** [comment]
I did not get to the variable editor as these bugs appeared before I could get there :)

I recall you saying this might occur!

**Comment by yesoreyeram** [comment]
I can reproduce the issue and make a fix here https://github.com/yesoreyeram/grafana-infinity-datasource/releases/tag/v0.8.0-dev.7 The fix worked for me. But want to make sure nothing else is broken.

**Comment by yesoreyeram** [comment]
+ @txjmb in case if you want to confirm.

**Comment by gdiana** [comment]
Ok I tried the new fix out and there are still some issues.

In terms of the table (ENERGY AND DEMAND) it now accepts the number format but numbers are returned with xxxx : , a colon on the end. I fixed this by using a format override and all appears ok. Not sure why it adds the : 
See below

![image](https://user-images.githubusercontent.com/13445152/141020372-72c50801-0e40-490a-b09a-c3fe30c8316e.png)

Then I tried same with time series plot and it did not work and the line plot did not work as no data was returned as shown below.

![image](https://user-images.githubusercontent.com/13445152/141020613-135794be-7f1f-4546-8c81-b0c02a2e95c9.png)

So I then changed the format to TABLE and it worked as shown below.

![image](https://user-images.githubusercontent.com/13445152/141020736-6fed42f0-ada1-4c6f-9109-41bc6a536017.png)

It also works with format as DATA FRAME.

So one problem appears to be when using the Times Series FORMAT, then no data is returned and no plot shows.

When using TABLE format it recognizes numbers but appends a colon (:)  12345 :

Also if I try use a variable query the dashboard takes ages to load, so much so I deleted it. It would be nice to also include an option not to load the variable query either at startup or on time range change!

**Comment by gdiana** [comment]
Oh I have now noticed that the as the TIME SERIES format does not work that the series overrides do not work and that on the legend it prints the letter of the query first for each legend variable.

Thus whereas before it would show MD max 300 it now shows B MD max 300.

So it appears that the QUERY NAME appearing before each legend item is a bug that has crept in somewhere as well.

![image](https://user-images.githubusercontent.com/13445152/141056775-534521c2-61c2-43f0-9611-65f964356763.png)

To highlight this issue I just changed the name of the A query to RRRRRRRRRRR and the effect is shown below

![image](https://user-images.githubusercontent.com/13445152/141057068-99aae6b7-8b22-4036-ba06-622b1ce102db.png)

If I change the format to TIME SERIES this bug does not appear but no data is displayed either as shown below

![image](https://user-images.githubusercontent.com/13445152/141057258-0c4d6f83-2186-44c3-a4a0-138fe73a9055.png)

The purple dot is the maximum point but the query is set to TABLE format.

So it appears these minor bug issues have crept in and are related to legend variables now being preceded by the query name but not in TIME SERIES format but which does not show any data either. Both TABLE and DATA FRAME formats display the same behaviour.
